In [2]:
import pandas as pd
url = " https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df.isna().sum()

# Check for missing values in the entire DataFrame
missing_values = df.isnull().sum()

print("Missing values in each column:")
print(missing_values)

# Check if there are any missing values at all
if df.isnull().values.any():
    print("\nThe dataset contains missing values.")
else:
    print("\nThe dataset has no missing values.")


Missing values in each column:
survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

The dataset contains missing values.


#2

In [8]:
import pandas as pd


df = pd.DataFrame(df)

# Get the number of rows and columns
num_rows, num_columns = df.shape

print(f"The DataFrame has {num_rows} rows and {num_columns} columns.")

The DataFrame has 891 rows and 15 columns.


#2 "observation" : observation refers to instances in the dataset, where values were measured/observed for all variables. Observations are often the "rows" of a dataframe
In the example about, each observation is each person who were abaord titanic when it sunk

"variables" : variables are what are being measured/observed, they are usually the columns of a dataframe
In the example, the variables are "survived", "pclass", "sex", "age", etc


In [3]:
print("Shape of the DataFrame:")
print(df.shape)  # This will print a tuple (number of rows, number of columns)

# Use df.describe() to get summary statistics of numerical columns
print("\nSummary statistics with df.describe():")
print(df.describe())

# Use df.describe(include=['object']) to get summary statistics of categorical columns
print("\nSummary statistics of categorical columns with df.describe(include=['object']):")
print(df.describe(include=['object']))

Shape of the DataFrame:
(891, 15)

Summary statistics with df.describe():
         survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

Summary statistics of categorical columns with df.describe(include=['object']):
         sex embarked  class  who deck  embark_town alive
count    891      889    891  891  203          889   891
unique     2        3      

#4 df.shape only gives you how many row/columns while df.describe() gives details about what's in each column, giving you the count, mean, std, min... max of the column. 
But df.describe will only be able to give you summary for numerical column
While df.shape gives you the full dimension of the data frame, df.describe() will skips the missing values, (as you can see under age/count), even tho there are 891 rows, df.describe only counted 714 under age. 
Also df.shape is an attribute that doesn't require Parentheses while df.describe() is a function that does require parentheses  

#5 attributes variables that are associated with an object, it is used to hold data about the object, and it is accessed by using dot notation without paraenthese
on the other hand
methods are actions that are assoicated with an object, it is used to perform certain operations on the object, it can be called also using dot notation but with paraenthese after it to include arguments

#6 count: the number of non missing value in the column
mean: the average of the non missing value in the column
std: the standard deviation of the non missing value in the column
min: the lowest value in the column
25%: 25th percentile of the data, or Q1
50%: 50th percentile of the data, or mediumn 
75%: 75th percentile of the data, or Q3
max: the highest value int he column


#7 
1. you might find "df.dropna()" a better method than "del df['col']" when you don't want to permanently delete an entire column
2. you might want to use "del df['col']" when you want to delete a specific column rather than just the ones with missing value
3. You might want to apply "del df['col']" before you use "df.dropna()" when you want to remove specific data from the dataframe before you removing the missing values, since having missing values in the column may affect whcih row/column "df.dropna" with drop


In [17]:
print(df)
del df['deck']
del df['age']
df.dropna()
print(df)

     survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alo

In [ ]:
missing_values = df.isnull().sum()

print("Missing values in each column:")
print(missing_values)

# Check if there are any missing values at all
if df.isnull().values.any():
    print("\nThe dataset contains missing values.")
else:
    print("\nThe dataset has no missing values.")

#8 df.groupby('col1')['col2'].describe() will create a new dataframe for each unique value in 'col1' and assign all the values in 'col2' to the new groups, then it would run the usual .describe() for each of the new group, giving you the count, mean, std, min, etc

I will do an example below, using 'deck' and 'fare'

In [3]:
df.groupby('deck')['fare'].describe()

count        mean         std      min      25%       50%        75%  \
deck                                                                         
A      15.0   39.623887   17.975333   0.0000  30.8479  35.50000   50.24790   
B      47.0  113.505764  109.301500   0.0000  57.0000  80.00000  120.00000   
C      59.0  100.151341   70.225588  26.5500  42.5021  83.47500  143.59165   
D      33.0   57.244576   29.592832  12.8750  30.0000  53.10000   77.28750   
E      32.0   46.026694   32.608315   8.0500  26.1125  45.18125   56.15730   
F      13.0   18.696792   11.728217   7.6500   7.7500  13.00000   26.00000   
G       4.0   13.581250    3.601222  10.4625  10.4625  13.58125   16.70000   

           max  
deck            
A      81.8583  
B     512.3292  
C     263.0000  
D     113.2750  
E     134.5000  
F      39.0000  
G      16.7000

Rather than showing just how much missing data there are, the count also shows the amount of elements in each new group

#8.3

In [1]:
df

NameError: name 'df' is not defined

In [1]:
import pandas as pd
url = " https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanics.csv"
df = pd.read_csv(url)

HTTPError: HTTP Error 404: Not Found

In [4]:
import pandas as pd
url = " https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
DF.groupby("col1")["col2"].describe()

KeyError: 'col1'

In [8]:
import pandas as pd
url = " https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df.group_by("col1")["col2"].describe()

AttributeError: 'DataFrame' object has no attribute 'group_by'

In [9]:
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df.groupby("deck")["Fare"].describe()

KeyError: 'Column not found: Fare'

In [10]:
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df.groupby("deck")[fare].describe()


NameError: name 'fare' is not defined

After experimenting with both google and chatgpt to troubleshoot the codes above, I find that using chat was a lot faster at pin pointing exactly where the code when wrong and it also provided the correct code for me to try them out. Chat gpt also gives you explanation on what was the mistake and why the code needs to be changed

Using google is similar, but I couldn't just copy and paste my code into the brower, and it requires a bit of thinking and scrolling through the error messages. But often, the error message will give you a clear idea on what and where the code went wrong. Instead of giving you the correct code, searching on google will require you to understand what the error message is saying and how to fix it. It may take some work but it gives you a better understanding so you can better handle it next time. 

Overall, both methods are great, but chatgpt is a lot faster and straight forward. And since it has history of your code and exact dataset you are working with. It can come up with code for your situation while google will only teach you the ways to fix the code. 



#9 mostly

ChatGPT links: 
frist chat: https://chatgpt.com/share/66e37ce7-134c-8010-a722-b2581a55ca9a
Second chat: https://chatgpt.com/share/66e37cb8-33e4-8010-b94c-1b91ab15f7da